runtime/fail_nat.c

```diff
@@ -195,20 +201,20 @@ CAMLexport value caml_raise_if_exception(value res)
    do a GC before the exception is raised (lack of stack descriptors
    for the ccall to [caml_array_bound_error]).  */
 
-static const value * caml_array_bound_error_exn = NULL;
-
 void caml_array_bound_error(void)
 {
-  if (caml_array_bound_error_exn == NULL) {
-    caml_array_bound_error_exn =
-      caml_named_value("Pervasives.array_bound_error");
-    if (caml_array_bound_error_exn == NULL) {
+  static atomic_uintnat exn_cache = ATOMIC_UINTNAT_INIT(0);
+  const value* exn = (const value*)atomic_load_acq(&exn_cache);
+  if (!exn) {
+    exn = caml_named_value("Pervasives.array_bound_error");
+    if (!exn) {
       fprintf(stderr, "Fatal error: exception "
-                      "Invalid_argument(\"index out of bounds\")\n");
+        "Invalid_argument(\"index out of bounds\")\n");
       exit(2);
```

Comment:
To do: change this into a call to `caml_fatal_error`.

Reply:
Although cf. https://github.com/ocaml/ocaml/pull/10726#issuecomment-981702733

Reply:
Yes, this is an odd choice of mine that is not completely illogical and goes way back, so let's not discuss it now.